File: packages/ironauth-sdk/snippets/verify-log-stream.mjs
// SPDX-License-Identifier: MIT OR Apache-2.0

/**
 * Verify a shipped IronAuth log-stream batch (issue #110 criterion 5).
 *
 * COPY THIS FILE. It imports nothing and uses only WebCrypto, so it drops into a SIEM
 * forwarder, a Lambda that reads the S3 sink's objects, or a Node handler behind the HTTP
 * sink, unchanged.
 *
 * ## What TLS does not tell you, and this does
 *
 * TLS protects the hop. Once a batch has landed in an object store, a forwarder or a log
 * index, it says nothing about where the batch came from or whether you already applied it.
 * The AWS SigV4 signing on the S3 sink is transport authentication to S3 -- discarded the
 * moment the object is written, and absent for the HTTP, Datadog and Splunk sinks.
 *
 * This answers the three questions a SIEM actually has:
 *
 *   1. AUTHENTICITY  -- did this come from the deployment, or from anyone who can write to
 *                       the bucket?
 *   2. ORDERING      -- is this the batch after the one I last verified?
 *   3. REPLAY        -- have I applied this already under another name?
 *
 * Ordering and replay are answered by the CURSOR POSITION, which is inside the signature
 * rather than beside it. The shipper advances the cursor only on success and only to what was
 * accepted, so positions are monotonic per stream. Keep the last position you verified and:
 *
 *   - a position at or below it is a REPLAY.
 *
 * Replay is answerable with what a batch carries. `verifyBatch` refuses without a position
 * (`missing-position`) and returns the position it verified rather than a bare boolean, so a
 * caller can record it.
 *
 * GAP DETECTION IS NOT AVAILABLE, and this file used to imply it was. The position is a
 * wall-clock microsecond timestamp of the last row in the batch, not a counter, so there is
 * no "next expected" value to compare against: a consumer can prove it has not seen a
 * position before, and cannot prove it has missed nothing in between. Detecting gaps would
 * need the batch to carry its START position as well, so positions chain.
 *
 * ## Where the position comes from
 *
 * The `x-ironauth-log-position` header, as `<stream id> <cursor sequence> <cursor id>`, space
 * separated. The S3 sink carries the same value as the `x-amz-meta-ironauth-log-position`
 * object metadata key, since an object has no headers once written.
 *
 *     const [streamId, cursorSequence, cursorId] = positionHeader.split(' ');
 *
 * Split positionally: all three are opaque identifiers, which is why the separator is a space
 * rather than a character that could occur inside one.
 *
 * This is worth stating here because it was missing for a while. The header was defined and
 * documented in the shipper and sent by nothing, so this file required three values the wire
 * never carried: a published verifier that could not be fed. See
 * `docs/log-stream-verification.md` for the full wire description.
 *
 * ## Conformance
 *
 * This file and the Rust signer are kept honest by the same corpus
 * (`../vectors/log-stream-vectors.json`), which is GENERATED from the signer. Two
 * implementations that agree on seven vectors -- one accepting and six adversarial -- are two
 * implementations that agree.
 */

/** The canonical-form version this file understands. */
export const CANONICAL_VERSION = 'ironauth-log-stream-v1';

const encoder = new TextEncoder();

/** Lowercase hex, matching the signer's encoding exactly. */
function hex(buffer) {
  return [...new Uint8Array(buffer)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

/**
 * Rebuild the string the signature covers.
 *
 * Built here rather than trusted from the wire. A consumer handed a "canonical string" by the
 * sender is verifying that the sender can sign its own claims, which is not a check.
 */
export async function canonicalString({
  streamId,
  cursorSequence,
  cursorId,
  eventCount,
  eventsJson,
}) {
  const digest = hex(await crypto.subtle.digest('SHA-256', encoder.encode(eventsJson)));
  return [
    CANONICAL_VERSION,
    streamId,
    String(cursorSequence),
    cursorId,
    String(eventCount),
    digest,
  ].join('\n');
}

/**
 * Verify one batch. Returns `{ ok, reason, position }`.
 *
 * `lastVerifiedSequence` is optional ONLY so a first-ever batch can be verified. Pass it
 * every time after that: without it this checks authenticity and integrity but cannot check
 * ordering or replay, and a caller who never passes it has a verifier that a replayed batch
 * sails straight through.
 */
export async function verifyBatch({
  key,
  signature,
  streamId,
  cursorSequence,
  cursorId,
  eventCount,
  eventsJson,
  lastVerifiedSequence = null,
}) {
  // A MISSING POSITION IS ITS OWN REFUSAL, and it has to be, because the alternative is
  // silent. Without this, an omitted `cursorSequence` builds the canonical string with the
  // literal `undefined` in it, the HMAC does not match, and the caller is told
  // `bad-signature`: an integration bug reported as an attack. This file previously said it
  // "refuses to be used without a position" while doing exactly that, which was measured.
  if (
    typeof streamId !== 'string' || streamId.length === 0 ||
    !Number.isInteger(cursorSequence) ||
    typeof cursorId !== 'string' || cursorId.length === 0
  ) {
    return { ok: false, reason: 'missing-position', position: null };
  }

  const canonical = await canonicalString({
    streamId,
    cursorSequence,
    cursorId,
    eventCount,
    eventsJson,
  });

  // Malformed input is a REFUSAL, never a throw. A verifier that throws on a bad signature
  // hands anyone who can reach it a denial of service in place of a `false`.
  if (typeof signature !== 'string' || !/^[0-9a-fA-F]*$/.test(signature) || signature.length % 2) {
    return { ok: false, reason: 'malformed-signature', position: null };
  }

  const imported = await crypto.subtle.importKey(
    'raw',
    typeof key === 'string' ? encoder.encode(key) : key,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['verify'],
  );
  const bytes = new Uint8Array(
    signature.match(/../g)?.map((pair) => Number.parseInt(pair, 16)) ?? [],
  );

  // `crypto.subtle.verify` is constant time. Comparing hex strings with === would leak the
  // position of the first difference to anyone who can submit candidates and time the answer.
  const ok = await crypto.subtle.verify('HMAC', imported, bytes, encoder.encode(canonical));
  if (!ok) {
    return { ok: false, reason: 'bad-signature', position: null };
  }

  if (lastVerifiedSequence !== null) {
    if (cursorSequence <= lastVerifiedSequence) {
      return { ok: false, reason: 'replay', position: cursorSequence };
    }
  }

  return { ok: true, reason: null, position: cursorSequence };
}

File: examples/verify-log-stream.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT OR Apache-2.0
"""Verify a signed IronAuth log-stream batch (issue #110 criterion 5).

This is the SAMPLE CONSUMER. It is deliberately dependency-free and short enough to read
in one sitting, because a SIEM operator has to be able to satisfy themselves that it does
what it says before they trust it with their audit trail.

It answers the three questions TLS cannot answer once a batch has landed in an object
store, a forwarder or a log index:

  authenticity  did this batch come from the deployment that holds the signing secret
  ordering      is this the batch that follows the one I last verified
  replay        have I already applied this batch under a different name

WHAT IS SIGNED

  ironauth-log-stream-v1
  <stream id>
  <cursor sequence>
  <cursor id>
  <event count>
  <SHA-256 of the serialized events, lowercase hex>

joined with newlines, and HMAC-SHA256'd with the stream's signing secret. The digest is
over the events array exactly as IronAuth serialized it, which is what the HTTP and S3
sinks transmit verbatim.

VENDOR SINKS RESHAPE THE PAYLOAD, and this is the one place a consumer has to do work:
the Datadog sink wraps each event as {"ddsource","service","message":<event>} and the
Splunk HEC sink wraps each as {"sourcetype","event":<event>}, newline delimited. Pass
--sink datadog or --sink splunk and this unwraps them back to the array that was signed.
The unwrapping is deterministic, so verification is possible on every sink; it is not
guesswork.

USAGE

  verify-log-stream.py --secret-file key.bin --body batch.json \
      --position "$X_IRONAUTH_LOG_POSITION" --signature "$X_IRONAUTH_LOG_SIGNATURE"

exits 0 when the batch verifies, 1 when it does not, and prints the reason either way.
"""

import argparse
import hashlib
import hmac
import json
import sys

CANONICAL_VERSION = "ironauth-log-stream-v1"


def events_from(body: str, sink: str) -> str:
    """The events array as IronAuth serialized it, whatever the sink wrapped it in."""
    if sink == "raw":
        # HTTP and S3 transmit the signed bytes verbatim. Re-serializing would risk
        # changing separators or key order, so the body is used exactly as received.
        return body
    if sink == "datadog":
        return json.dumps([entry["message"] for entry in json.loads(body)], separators=(",", ":"))
    if sink == "splunk":
        events = [json.loads(line)["event"] for line in body.splitlines() if line.strip()]
        return json.dumps(events, separators=(",", ":"))
    raise SystemExit(f"unknown sink {sink!r}")


def canonical_string(stream_id: str, sequence: int, cursor_id: str, events_json: str) -> str:
    count = len(json.loads(events_json))
    digest = hashlib.sha256(events_json.encode("utf-8")).hexdigest()
    return "\n".join(
        [CANONICAL_VERSION, stream_id, str(sequence), cursor_id, str(count), digest]
    )


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--secret-file", required=True, help="the stream's signing secret, raw bytes")
    parser.add_argument("--body", required=True, help="the batch body as received")
    parser.add_argument("--position", required=True, help="x-ironauth-log-position header value")
    parser.add_argument("--signature", required=True, help="x-ironauth-log-signature header value")
    parser.add_argument("--sink", default="raw", choices=["raw", "datadog", "splunk"])
    parser.add_argument("--last-sequence", type=int, default=None,
                        help="the sequence you last verified, to check ordering and replay")
    args = parser.parse_args()

    with open(args.secret_file, "rb") as handle:
        key = handle.read()
    with open(args.body, "r", encoding="utf-8") as handle:
        body = handle.read()

    # "<stream id> <sequence> <cursor id>", split positionally: all three are opaque ids.
    parts = args.position.split(" ")
    if len(parts) != 3:
        print(f"malformed position header: {args.position!r}", file=sys.stderr)
        return 1
    stream_id, sequence, cursor_id = parts[0], int(parts[1]), parts[2]

    canonical = canonical_string(stream_id, sequence, cursor_id, events_from(body, args.sink))
    expected = hmac.new(key, canonical.encode("utf-8"), hashlib.sha256).hexdigest()
    if not hmac.compare_digest(expected, args.signature):
        print("SIGNATURE MISMATCH: this batch was not signed by that secret, or it was altered",
              file=sys.stderr)
        return 1

    # Ordering and replay are the reason the position is signed rather than merely sent.
    if args.last_sequence is not None:
        if sequence <= args.last_sequence:
            print(f"REPLAY: sequence {sequence} is not after {args.last_sequence}", file=sys.stderr)
            return 1

    print(f"verified: stream {stream_id} at sequence {sequence}, cursor {cursor_id}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
